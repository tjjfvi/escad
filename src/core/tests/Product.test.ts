
import {
  ArrayProduct,
  ArrayProductType,
  createLeafProductUtils,
  Id,
  LeafProduct,
  Product,
  ProductType,
  TupleProduct,
  TupleProductType,
  UnknownProduct,
} from "../src"

const productAId = Id.create(__filename, "@escad/core", "LeafProduct", "ProductA")
const productBId = Id.create(__filename, "@escad/core", "LeafProduct", "ProductB")
const productCId = Id.create(__filename, "@escad/core", "LeafProduct", "ProductC")
interface ProductA extends LeafProduct { readonly type: typeof productAId }
interface ProductB extends LeafProduct { readonly type: typeof productBId }
interface ProductC extends LeafProduct { readonly type: typeof productCId }
const ProductA = {
  create: (): ProductA => ({ type: productAId }),
  id: productAId,
  ...createLeafProductUtils<ProductA, "ProductA">(productAId, "ProductA"),
}
const ProductB = {
  create: (): ProductB => ({ type: productBId }),
  id: productBId,
  ...createLeafProductUtils<ProductB, "ProductB">(productBId, "ProductB"),
}
const ProductC = {
  create: (): ProductC => ({ type: productCId }),
  id: productCId,
  ...createLeafProductUtils<ProductC, "ProductC">(productCId, "ProductC"),
}

describe("createLeafProductUtils", () => {
  test("Consistent keys", () => {
    expect(Object.keys(ProductA)).toMatchSnapshot()
  })
  test(".productType", () => {
    expect(ProductA.productType).toMatchSnapshot()
  })
  describe(".isProductA", () => {
    const productA = ProductA.create()
    const productB = ProductB.create()
    test("Matches productA", () => {
      expect(ProductA.isProductA(productA)).toEqual(true)
    })
    test("Doesn't match productB", () => {
      expect(ProductA.isProductA(productB)).toEqual(false)
    })
    // Type guard
    const product: ProductA | null = ProductA.isProductA(productB) ? productB : null
    product
  })
})

const createProduct = () => TupleProduct.create([
  ProductA.create(),
  TupleProduct.create([]),
  ArrayProduct.create([ProductA.create(), ProductA.create()]),
  TupleProduct.create([ProductA.create(), ProductB.create(), ProductC.create()]),
  UnknownProduct.create(ProductB.create()),
])

// eslint-disable-next-line func-call-spacing
describe.each<readonly [string, () => Product]>([
  ["ProductA", () => ProductA.create()],
  ["TupleProduct<[]>", () => TupleProduct.create([])],
  ["TupleProduct<[ProductA, ProductB]>", () => TupleProduct.create([ProductA.create(), ProductB.create()])],
  ["ArrayProduct<ProductA>", () => ArrayProduct.create([ProductA.create(), ProductA.create()])],
  ["UnknownProduct", () => UnknownProduct.create(ProductB.create())],
  ["Complex", () => createProduct()],
] as const)("%s", (_, createProduct) => {
  test(".create", () => {
    expect(createProduct()).toMatchSnapshot()
  })
  test("Product.isProduct", () => {
    expect(Product.isProduct(createProduct())).toBe(true)
  })
  test("Product.getProductType", () => {
    expect(Product.getProductType(createProduct())).toMatchSnapshot()
  })
  test("Product.isProduct with correct ProductType", () => {
    expect(Product.isProduct(createProduct(), Product.getProductType(createProduct()))).toBe(true)
  })
  describe("Product.isProduct with wrong ProductType", () => {
    test.each([
      ["ProductC", ProductC],
      ["ArrayProduct<ProductC>", ArrayProductType.create(ProductC)],
      ["TupleProduct<[ProductA, ProductC]>", TupleProductType.create([ProductA, ProductC])],
    ] as const)("%s", (_, productType) => {
      expect(Product.isProduct(createProduct(), ProductType.fromProductTypeish<Product>(productType))).toBe(false)
    })
  })
})

describe("Product.getProductType", () => {
  test("Throw on invalid product", () => {
    expect(() => Product.getProductType(null as never)).toThrowErrorMatchingSnapshot()
  })
})

describe("ArrayProduct.create", () => {
  test("Throws on []", () => {
    expect(() => ArrayProduct.create([])).toThrowErrorMatchingSnapshot()
  })
})
