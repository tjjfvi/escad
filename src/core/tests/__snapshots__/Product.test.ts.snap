// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayProduct.create Throws on [] 1`] = `"ArrayProducts cannot be empty"`;

exports[`ArrayProduct<ProductA> .create 1`] = `
Object {
  "children": Array [
    Object {
      "type": "@escad/core/LeafProduct/ProductA",
    },
    Object {
      "type": "@escad/core/LeafProduct/ProductA",
    },
  ],
  "type": "ArrayProduct",
}
`;

exports[`ArrayProduct<ProductA> Product.getProductType 1`] = `
Object {
  "elementType": Object {
    "id": "@escad/core/LeafProduct/ProductA",
    "type": "LeafProductType",
  },
  "type": "ArrayProductType",
}
`;

exports[`Complex .create 1`] = `
Object {
  "children": Array [
    Object {
      "type": "@escad/core/LeafProduct/ProductA",
    },
    Object {
      "children": Array [],
      "type": "TupleProduct",
    },
    Object {
      "children": Array [
        Object {
          "type": "@escad/core/LeafProduct/ProductA",
        },
        Object {
          "type": "@escad/core/LeafProduct/ProductA",
        },
      ],
      "type": "ArrayProduct",
    },
    Object {
      "children": Array [
        Object {
          "type": "@escad/core/LeafProduct/ProductA",
        },
        Object {
          "type": "@escad/core/LeafProduct/ProductB",
        },
        Object {
          "type": "@escad/core/LeafProduct/ProductC",
        },
      ],
      "type": "TupleProduct",
    },
    Object {
      "product": Object {
        "type": "@escad/core/LeafProduct/ProductB",
      },
      "type": "UnknownProduct",
    },
  ],
  "type": "TupleProduct",
}
`;

exports[`Complex Product.getProductType 1`] = `
Object {
  "elementTypes": Array [
    Object {
      "id": "@escad/core/LeafProduct/ProductA",
      "type": "LeafProductType",
    },
    Object {
      "elementTypes": Array [],
      "type": "TupleProductType",
    },
    Object {
      "elementType": Object {
        "id": "@escad/core/LeafProduct/ProductA",
        "type": "LeafProductType",
      },
      "type": "ArrayProductType",
    },
    Object {
      "elementTypes": Array [
        Object {
          "id": "@escad/core/LeafProduct/ProductA",
          "type": "LeafProductType",
        },
        Object {
          "id": "@escad/core/LeafProduct/ProductB",
          "type": "LeafProductType",
        },
        Object {
          "id": "@escad/core/LeafProduct/ProductC",
          "type": "LeafProductType",
        },
      ],
      "type": "TupleProductType",
    },
    Object {
      "type": "UnknownProductType",
    },
  ],
  "type": "TupleProductType",
}
`;

exports[`Product.getProductType Throw on invalid product 1`] = `"Invalid product passed to Product.getProductType"`;

exports[`ProductA .create 1`] = `
Object {
  "type": "@escad/core/LeafProduct/ProductA",
}
`;

exports[`ProductA Product.getProductType 1`] = `
Object {
  "id": "@escad/core/LeafProduct/ProductA",
  "type": "LeafProductType",
}
`;

exports[`TupleProduct<[]> .create 1`] = `
Object {
  "children": Array [],
  "type": "TupleProduct",
}
`;

exports[`TupleProduct<[]> Product.getProductType 1`] = `
Object {
  "elementTypes": Array [],
  "type": "TupleProductType",
}
`;

exports[`TupleProduct<[ProductA, ProductB]> .create 1`] = `
Object {
  "children": Array [
    Object {
      "type": "@escad/core/LeafProduct/ProductA",
    },
    Object {
      "type": "@escad/core/LeafProduct/ProductB",
    },
  ],
  "type": "TupleProduct",
}
`;

exports[`TupleProduct<[ProductA, ProductB]> Product.getProductType 1`] = `
Object {
  "elementTypes": Array [
    Object {
      "id": "@escad/core/LeafProduct/ProductA",
      "type": "LeafProductType",
    },
    Object {
      "id": "@escad/core/LeafProduct/ProductB",
      "type": "LeafProductType",
    },
  ],
  "type": "TupleProductType",
}
`;

exports[`UnknownProduct .create 1`] = `
Object {
  "product": Object {
    "type": "@escad/core/LeafProduct/ProductB",
  },
  "type": "UnknownProduct",
}
`;

exports[`UnknownProduct Product.getProductType 1`] = `
Object {
  "type": "UnknownProductType",
}
`;

exports[`createLeafProductUtils .productType 1`] = `
Object {
  "id": "@escad/core/LeafProduct/ProductA",
  "type": "LeafProductType",
}
`;

exports[`createLeafProductUtils Consistent keys 1`] = `
Array [
  "create",
  "id",
  "productType",
  "isProductA",
  "convert",
]
`;
